### PR TITLE
Support unix domain socket

### LINKED ISSUE #24095: Consider making ShardingSphere-Proxy support Unix Domain Socket
### Describe the feature you would like.

The ShardingSphere-Proxy support TCP/IP only for now.
If ShardingSphere-Proxy supported UDS:
- Improve local access network performance.
- Client could connect to Proxy without password.

### PR DESCRIPTION
**Pull request by sunkai-cai**
Fixes #24095.

 ShardingSphere-Proxy supported UDS。

Changes proposed in this pull request:
  -  In  ShardingSphere-Proxy，add supported UDS by the `socket path` paramet.
  -  In  start.sh，add `-s` of paramet  to setting  `socket path`.

---

Before committing this PR, I'm sure that I have checked the following options:
- [ ] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [ ] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [ ] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added corresponding unit tests for my changes.

### CODE REVIEW

File: proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java [hunk 1]
```diff
@@ -97,4 +118,13 @@ private String paddingWithSlash(final String pathArg) {
         }
         return result.toString();
     }
+    
+    private boolean isValidPath(final String path) {
+        try {
+            Paths.get(path);
+        } catch (InvalidPathException | NullPointerException ex) {
+            return false;
```
Comment (TeslaCN) [inline proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java:126]: My understanding is that if user specified an invalid path, the socket will not be present and no error will occur. Am I right?

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            future.channel().closeFuture().sync();
+        } catch (final InterruptedException ignored) {
+        } finally {
+            close();
+        }
+    }
+    
+    private ChannelFuture startDomainSocket(final String socketPath) throws InterruptedException {
+        log.info("DomainSocket Starting {}", socketPath);
+        createEventLoopGroup();
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java:64]: Could we reuse the EventLoopGroup?

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        ServerBootstrap bootstrap = new ServerBootstrap();
+        initServerBootstrap(bootstrap, new DomainSocketAddress(socketPath));
+        Runtime.getRuntime().addShutdownHook(new Thread(this::close));
+        return bootstrap.bind().sync();
+    }
+    
+    private void createEventLoopGroup() {
+        bossGroup = Epoll.isAvailable() ? new EpollEventLoopGroup(1) : new NioEventLoopGroup(1);
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java:72]: If the Epoll was not available, how EpollServerDomainSocketChannel work?

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.shardingsphere.proxy.backend.context.ProxyContext;
+import org.apache.shardingsphere.proxy.frontend.netty.ServerHandlerInitializer;
+import org.apache.shardingsphere.proxy.frontend.protocol.FrontDatabaseProtocolTypeFactory;
+
+/**
+ * UnixDomainSocketServer.
+ */
+@Slf4j
+@RequiredArgsConstructor
+public final class UnixDomainSocketServer extends Thread {
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java:43]: Could we move the ServerBootstrap to class `ShardingSphereProxy`? UDS is just a way to connect Proxy.

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java [hunk 5]
```diff
@@ -100,6 +98,7 @@ private EventLoopGroup getWorkerGroup() {
     
     private void initServerBootstrap(final ServerBootstrap bootstrap) {
         Integer backLog = ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData().getProps().<Integer>getValue(ConfigurationPropertyKey.PROXY_NETTY_BACKLOG);
+
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:101]: The blank line could be removed.

File: proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java (hunk 1 above)
Comment (sunkai-cai) [reply proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java:126]: yes, it was designed that way.

Comment (sunkai-cai) [reply proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java:126]: of course, we can throw a error in this.

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        bossGroup = Epoll.isAvailable() ? new EpollEventLoopGroup(1) : new NioEventLoopGroup(1);
+        workerGroup = Epoll.isAvailable() ? new EpollEventLoopGroup() : new NioEventLoopGroup();
+    }
+    
+    private void initServerBootstrap(final ServerBootstrap bootstrap, final DomainSocketAddress localDomainSocketAddress) {
+        Integer backLog = ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData().getProps().<Integer>getValue(ConfigurationPropertyKey.PROXY_NETTY_BACKLOG);
+        bootstrap.group(bossGroup, workerGroup)
+                .channel(EpollServerDomainSocketChannel.class)
+                .localAddress(localDomainSocketAddress)
+                .option(ChannelOption.SO_BACKLOG, backLog)
```
[the remark this answers is not in the document]
Comment (sunkai-cai) [reply proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/UnixDomainSocketServer.java:81]: ok

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java (hunk 5 above)
Comment (sunkai-cai) [reply proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:101]: ok

File: proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java [hunk 7]
```diff
@@ -56,6 +58,7 @@
 /**
  * Authentication engine for MySQL.
  */
+@Slf4j
```
Comment (TeslaCN) [inline proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java:61]: Please remove unused annotation.

File: proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java [hunk 8]
```diff
@@ -147,6 +150,10 @@ private boolean authorizeDatabase(final AuthorityRule rule, final Grantee grante
     }
     
     private String getHostAddress(final ChannelHandlerContext context) {
+        if (context.channel() instanceof EpollDomainSocketChannel) {
+            return context.channel().parent().localAddress().toString();
+        }
+
```
Comment (TeslaCN) [inline proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java:156]: Please remove the blank line.

File: proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java (hunk 1 above)
Comment (TeslaCN) [reply proxy/bootstrap/src/main/java/org/apache/shardingsphere/proxy/arguments/BootstrapArguments.java:129]: I prefer throwing an exception if path was invalid.

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java [hunk 9]
```diff
@@ -89,8 +123,12 @@ private void accept(final List<ChannelFuture> futures) throws InterruptedExcepti
     }
     
     private void createEventLoopGroup() {
-        bossGroup = Epoll.isAvailable() ? new EpollEventLoopGroup(1) : new NioEventLoopGroup(1);
-        workerGroup = getWorkerGroup();
+        if (null == bossGroup) {
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:126]: EventLoopGroup could be created once and used in both ServerBootstrap. Please avoid null check here.

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java (hunk 8 above)
Comment (sunkai-cai) [reply proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java:156]: ok

File: proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java (hunk 7 above)
Comment (sunkai-cai) [reply proxy/frontend/type/mysql/src/main/java/org/apache/shardingsphere/proxy/frontend/mysql/authentication/MySQLAuthenticationEngine.java:61]: ok

**Comment by TeslaCN** [comment]
Please add tests to cover your changes.

**Comment by sunkai-cai** [comment]
> Please add tests to cover your changes.

ok

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java [hunk 10]
```diff
@@ -111,4 +146,22 @@ private void initServerBootstrap(final ServerBootstrap bootstrap) {
...
+    private void initServerBootstrap(final ServerBootstrap bootstrap, final DomainSocketAddress localDomainSocketAddress) {
+        bootstrap.group(bossGroup, workerGroup)
+                .channel(EpollServerDomainSocketChannel.class)
+                .localAddress(localDomainSocketAddress)
+                .handler(new LoggingHandler(LogLevel.INFO))
+                .childHandler(new ServerHandlerInitializer(FrontDatabaseProtocolTypeFactory.getDatabaseType()));
+    }
+    
+    private void close() {
+        if (null != bossGroup) {
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:159]: Is this null check necessary?

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java [hunk 11]
```diff
@@ -63,24 +71,51 @@ public void start(final int port, final List<String> addresses) {
...
         initServerBootstrap(bootstrap);
-        
         List<ChannelFuture> futures = new ArrayList<>();
         for (String address : addresses) {
             futures.add(bootstrap.bind(address, port).sync());
         }
         return futures;
     }
     
+    private ChannelFuture startDomainSocket(final String socketPath) throws InterruptedException {
```
Comment (TeslaCN) [inline proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:113]: Does `InterruptedException` necessary?

Comment (sunkai-cai) [reply proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:113]: oh，it‘s not need。

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java [hunk 12]
```diff
@@ -63,24 +71,51 @@ public void start(final int port, final List<String> addresses) {
...
+            log.error("Epoll is unavailable, DomainSocket can't start.");
+            return;
+        }
+        try {
+            ChannelFuture future = startDomainSocket(socketPath);
+            future.addListener((ChannelFutureListener) futureParams -> {
+                if (futureParams.isSuccess()) {
+                    log.info("The listening address for DomainSocket is {}", socketPath);
+                } else {
+                    log.error("DomainSocket failed to start.");
```
[the remark this answers is not in the document]
Comment (sunkai-cai) [reply proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:94]: ok

File: proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java (hunk 10 above)
Comment (sunkai-cai) [reply proxy/frontend/core/src/main/java/org/apache/shardingsphere/proxy/frontend/ShardingSphereProxy.java:159]: we can remove it now。

**Review by TeslaCN — APPROVED** [review]
Thank you @sunkai-cai 
The codes are more clear now.

**Comment by sunkai-cai** [comment]
> Thank you @sunkai-cai The codes are more clear now.

thanks for your review for make it better.

**Pull request merged**